library/core/src/slice/iter/macros.rs

```diff
@@ -191,6 +191,32 @@ macro_rules! iterator {
                 self.next_back()
             }
 
+            #[inline]
+            fn fold<B, F>(self, init: B, mut f: F) -> B
+                where
+                    F: FnMut(B, Self::Item) -> B,
+            {
+                if is_empty!(self) {
+                    return init;
+                }
+                let mut acc = init;
+                let mut i = 0;
+                let len = len!(self);
+                loop {
+                    // SAFETY: the loop iterates `i in 0..len`, which always is in bounds of
+                    // the slice allocation
+                    acc = f(acc, unsafe { & $( $mut_ )? *self.ptr.add(i).as_ptr() });
```

Comment:
How much of the duplication here is essential?

Avoiding the `Option` loop seems entirely reasonable to me, but how does this approach compare to something much shorter?  For example, something like
```rust
for _ in 0..len!(self) {
    acc = f(acc, next_unchecked!(self));
}
```

Or the same with a `while (n --> 0)` loop or something, if the `Range` is not ok.

And if that's *not* sufficient, it would be nice to have some comments here about why it's written this way.

(Also, if overloading `fold` is worth it, I expect `rfold` should be overridden too.)

@rustbot author

Reply:
If you look at the PR history you'll see that I tried several approaches. A for-in-range loop was one of them. I even used `IndexRange` instead of `Range`.

> And if that's not sufficient, it would be nice to have some comments here about why it's written this way.

Ok, will do.

> (Also, if overloading fold is worth it, I expect rfold should be overridden too.)

Maybe, but that'd be optimizing two things at once which makes assessing perf more complicated.

Reply:
Added a comment.

@rustbot ready